api/v1/relay/relay.proto

```diff
@@ -0,0 +1,152 @@
+import public "flow/flow.proto";
+
+package relay;
+
+service HubbleRelay {
+    rpc GetEvents(GetEventsRequest) returns (stream GetEventsResponse) {}
+}
+
+// Here I didn't include "follow", "until", and "number". This request assumes follow,
+// and lets the client decide when to end the request, whether it's based on timestamp
+// or the number of responses received.
+message GetEventsRequest{
+    // EventType specifies which types of events to subscribe to. If unspecified, the
+    // request is subscribed to all the event types.
+    repeated EventType event_types = 1;
+    repeated EventFilter blacklist = 2;
+    repeated EventFilter whitelist = 3;
+    google.protobuf.Timestamp since = 4;
+}
+
+message GetEventsResponse{
+    string node = 1;
+    google.protobuf.Timestamp timestamp = 2;
+    oneof event {
+        flow.Flow flow = 3;
+        K8sNamespaceState k8s_namespace_state = 4;
+        ServiceState service_state = 5;
+        // It's unclear how we can define what it means for service links to get
+        // added/deleted. For now I'm deferring the decision for whether a particular
+        // link should be included in the service map to the client. Maybe the server
+        // can periodically send service_link events for active links to keep-alive,
+        // and the client can expire older links based on their timestamp.
+        ServiceLinkState  service_link_state = 6;
+    }
+}
+
+// Here I added different filter types for different events. Another option is to reuse
+// and extend FlowFilter to filter all the events, but I worry that it might become too
+// confusing in terms of figuring out which filter fields apply to which events.
+message EventFilter {
+    oneof filter {
+        flow.FlowFilter flow_filter = 2;
+        ServiceFilter service_filter = 3;
+        ServiceLinkFilter service_link_filter = 4;
+    }
+}
+
+enum EventType {
+    UNKNOWN_EVENT = 0;
+    FLOW = 1;
+    K8S_NAMESPACE_STATE = 2;
+    SERVICE_STATE = 3;
+    SERVICE_LINK_STATE = 4;
+}
+
+message K8sNamespace {
+    string id = 1;
+    string name = 2;
+    google.protobuf.Timestamp creation_timestamp = 3;
+}
+
+message K8sNamespaceState {
+    K8sNamespace namespace = 1;
+    StateChange type = 2;
+}
+
+// IP protocols. The values of enums do not correspond to actual IP protocol numbers.
+enum IPProtocol {
+    UNKNOWN_IP_PROTOCOL = 0;
+    TCP = 1;
+    UDP = 2;
+    ICMP_V4 = 3;
+    ICMP_V6 = 4;
+}
+
+// Come up with a better name for Service.
+message Service {
+    // An opaque ID that uniquely identifies the service.
+    string id = 1;
+    string name = 2;
+    string namespace = 3;
+    repeated string labels = 4;
+    repeated string dns_names = 5;
+    bool egress_policy_enforced = 6;
+    bool ingress_policy_enforced = 7;
+    // This is a string instead of boolean because Cilium reports status message. If it's
+    // successfully enabled, this string is set to "OK" I think.
+    string visibility_policy_status = 8;
```

Comment:
Yeah, if it's forwarded on directly it likely includes a diagnostic error if visibility is enabled and the visibility annotation is misconfigured.